src/main.rs

```diff
@@ -0,0 +1,291 @@
+                    self.peak_flag = 1;
+                }
+            } else {
+                self.threshold_ema += (fx - self.threshold_ema) * THRESHOLD_ALPHA_DN;
+                if self.state == 1 && self.timer >= PEAK_DELAY {
+                    self.update_hr(self.n - PEAK_DELAY);
+                    self.state = 0;
+                    self.timer = 0;
+                }
+            }
+            if self.state == 0 {
+                self.below_pts.push(x as i32 - self.threshold_ema as i32);
+            } else {
+                self.above_pts.push(x as i32 - self.threshold_ema as i32);
+            }
+        } else {
+            self.state = 0;
+            self.timer = 0;
+            self.wild_flag = 1;
+        }
+        self.n += 1;
+        self.timer += 1;
+        (self.n, x, self.peak_flag, self.state) // self.peak_flag)
+    }
+    fn update_hr(&mut self, start_n : usize) {
+        // Search for peak in above data
+        if self.above_pts.capacity() > 1 {
+            let mut above_max = 0i32;
+            let mut above_ix = 0usize;
+            for (i, val) in self.above_pts.iter().enumerate() {
+                if above_max < *val {
+                    above_max = *val;
+                    above_ix = i;
+                }
+            }
+            // Given when above_pts started, and above_ix, calc delta to last peak
+            let this_peak_n = start_n + above_ix;
+            let delta_n = this_peak_n - self.last_peak_n;
+            self.last_peak_n = this_peak_n;
+
+            // if delta > 200 && delta < 2000 {
+            self.hr = 60000f64 / delta_n as f64;
+
+            c5412::set_value(self.hr as u32);
+        }
+    }
+    fn hr(&self) -> f64 { self.hr }
+    fn above_below(&self) -> (i32, i32) { // 140us
+        let mut above = 0i32;
+        if self.above_pts.capacity() > 1 {
+            let i = self.above_pts.iter();
+            for ii in i {
+                if above < *ii {
+                    above = *ii;
+                }
+            }
+        }
+        let mut below = 0i32;
+        if self.below_pts.capacity() > 1 {
+            let i = self.below_pts.iter();
+            for ii in i {
+                if below > *ii {
+                    below = *ii;
+                }
+            }
+        }
+        (above, below)
+    }
+    fn new() -> Hr {
+        let yc = 32768i32;
+        Hr {
+            dc_ema: yc as f64,
+            lp_ema: yc as f64,
+            threshold_ema: yc as f64,
+            n : 0usize,
+            state : 0u8,
+            timer : 0usize,
+            peak_flag : 0u8,
+            wild_flag : 0u8,
+            above_pts : ConstGenericRingBuffer::<i32, ABOVE_SIZE>::new(),
+            below_pts : ConstGenericRingBuffer::<i32, BELOW_SIZE>::new(),
+            last_peak_n : 0usize,
+            hr : 0f64,
+        }
+    }
+    fn help(&self) -> (u32, u32) {
+        (self.dc_ema as u32, self.threshold_ema as u32)
+    }
+}
+
```

Comment:
Comment here about how threads need static scope on parameters, and the tricks needed to pass peripherals into threads

Reply:
done